code/modules/reagents/chemistry/machinery/chem_dispenser.dm

```diff
@@ -157,24 +147,50 @@
 			var/reagent = params["reagent"]
 			if(beaker && dispensable_reagents.Find(reagent))
 				var/datum/reagents/R = beaker.reagents
-				var/free = R.maximum_volume - R.total_volume
-				var/actual = min(amount, (cell.charge * powerefficiency)*10, free)
-
-				R.add_reagent(reagent, actual)
-				cell.use((actual / 10) / powerefficiency)
+				R.add_reagent(reagent, amount)
 				. = TRUE
 		if("remove")
 			var/amount = text2num(params["amount"])
 			if(beaker && amount in beaker.possible_transfer_amounts)
 				beaker.reagents.remove_all(amount)
 				. = TRUE
+		if("dispense_recipe")
+			var/recipe_to_use = params["recipe"]
+			var/list/chemicals_to_dispense = process_recipe_list(recipe_to_use)
+			for(var/r_id in chemicals_to_dispense) // i suppose you could edit the list locally before passing it
+				if(beaker && dispensable_reagents.Find(r_id)) // but since we verify we have the reagent, it'll be fine
+					var/amt = chemicals_to_dispense[r_id]
+					beaker.reagents.add_reagent(r_id, amt)
+		if("add_recipe")
+			var/name = stripped_input(usr,"Name","What do you want to name this recipe?", "15 Space Lube", MAX_NAME_LEN)
+			var/recipe = stripped_input(usr,"Recipe","Insert recipe with chem IDs", "oxygen=5;silicon=5;water=5")
+			if(name && recipe)
+				var/list/first_process = splittext(recipe, ";")
+				for(var/reagents in first_process)
+					var/list/fuck = splittext(reagents, "=")
+					if(dispensable_reagents.Find(fuck[1]))
+						continue
+					else
+						var/temp = fuck[1]
+						to_chat(usr, "[src] can't process [temp]!")
+						return
+				saved_recipes += list(list("recipe_name" = name, "contents" = recipe))
 		if("eject")
 			if(beaker)
 				beaker.forceMove(loc)
 				beaker = null
 				cut_overlays()
 				. = TRUE
 
+/obj/machinery/chem_dispenser/proc/process_recipe_list(var/fucking_hell)
```

Comment:
This form actually already exists, Byond Param strings, as such ```params2list()``` can replace this proc.

Reply:
Perfect

Reply:
Wait, does it support the delimiters I use

Reply:
Negative, this won't work for me. I tried params2list with my format, it doesn't work.

Reply:
it's ```"name1=value1&name2=value2&..."```, just swap your ; for &'s, and & is more natural for a user anyway.

Reply:
the thing is the docs say it should also support ;, which looks nicer

i'll give it a shot I guess